nucypher/policy/policies.py

```diff
@@ -728,56 +728,55 @@ def generate_policy_parameters(n: int,
         reservoir = self.alice.get_stakers_reservoir(duration=self.duration_periods,
                                                      without=handpicked_addresses)
+
+        quantity_remaining = self.n - len(handpicked_ursulas)
         if len(reservoir) < quantity_remaining:
-            error = f"Cannot create policy with {quantity_remaining} arrangements"
+            error = f"Cannot create policy with {self.n} arrangements"
             raise self.NotEnoughBlockchainUrsulas(error)
 
-        to_check = set(reservoir.draw(quantity_remaining))
+        # Handpicked Ursulas are not necessarily known
+        to_check = list(handpicked_ursulas) + reservoir.draw(quantity_remaining)
+        checked = []
 
         # Sample stakers in a loop and feed them to the learner to check
         # until we have enough in `selected_ursulas`.
 
         start_time = maya.now()
-        new_to_check = to_check
 
         while True:
 
             # Check if the sampled addresses are already known.
             # If we're lucky, we won't have to wait for the learner iteration to finish.
-            known = {x for x in to_check if x in self.alice.known_nodes}
-            to_check = to_check - known
-
-            known = random.sample(known, min(len(known), quantity_remaining)) # we only need so many
-            selected_ursulas.update([self.alice.known_nodes[address] for address in known])
-            quantity_remaining -= len(known)
+            checked = [x for x in to_check if x in self.alice.known_nodes]
+            to_check = [x for x in to_check if x not in self.alice.known_nodes]
 
-            if quantity_remaining == 0:
+            if len(checked) >= self.n:
                 break
-            else:
-                new_to_check = reservoir.draw_at_most(quantity_remaining)
-                to_check.update(new_to_check)
+
+            # Draw a little bit more nodes, if there are any
+            to_check += reservoir.draw_at_most(drawing_step)
 
             delta = maya.now() - start_time
             if delta.total_seconds() >= timeout:
                 still_checking = ', '.join(to_check)
+                quantity_remaining = self.n - len(checked)
                 raise RuntimeError(f"Timed out after {timeout} seconds; "
                                    f"need {quantity_remaining} more, still checking {still_checking}.")
 
-            self.alice.block_until_specific_nodes_are_known(to_check, learn_on_this_thread=discover_on_this_thread)
+            self.alice.block_until_specific_nodes_are_known(to_check,
+                                                            learn_on_this_thread=discover_on_this_thread,
+                                                            allow_missing=len(to_check),
+                                                            timeout=learner_timeout)
 
-        found_ursulas = list(selected_ursulas)
+        found_ursulas = [self.alice.known_nodes[address] for address in checked[:self.n]]
```

Comment:
Is it possible that selecting the first `n` of `checked` here can be introducing some bias? `checked` is a list, whose order comes from `to_check`, which in turn is ordered based on the handpicked ursulas and the reservoir output. At this point I don't know if there can be a bias or not, but maybe what you can do is do the shuffle first (which we already do anyway), and then trim the result to the desired length.

Reply:
We need to trim the list if some nodes we drew were unavailable, and we had to draw more nodes. The logic in the PR doesn't punish nodes for being unavailable if they managed to get online within the sampling timeout. What you propose will reduce chances of such nodes to be selected. I'm not against it really, just want to make the distinction clear. Also I don't think these situations will be too common - usually a node will be either available on the first try, or not available at all, right?